README.pod
=pod

=head1 NAME

Crypt::MCrypt - Perl interface for libmcrypt C library.

=head1 VERSION

version 0.07

=head1 SYNOPSIS

     use Crypt::MCrypt;
 
     my $iv = pack("H*","0000000000000000");
     my $key = pack("H*","1234567890123456" . "7890123456789012" . "1234567890123456");
     my $cipher_text = pack("H*","E9FF3161EE05ABC9" 
         . "7ea3cacb991318aa" 
         . "585379599b0eaabb" 
         . "c4e474ead1956f47" 
         . "6755f13f1af5235d");
     my $algorithm = "tripledes";
     my $mode = "cbc";
     my $obj = Crypt::MCrypt->new(
         algorithm => $algorithm, 
         mode      => $mode,
         key       => $key, 
         iv        => $iv,
     );
     my $plain_text = $obj->decrypt($cipher_text);
     print "\nPLAIN: $plain_text\n";
     print "\nPLAIN in hex: " . unpack("H*",$plain_text) . "\n";
     $cipher_text = $obj->encrypt($plain_text);
     print "\nCIPHER: $cipher_text\n";
     print "\nCIPHER in hex: " . unpack("H*",$cipher_text) . "\n";

=head1 DESCRIPTION

This is a perl interface to libmcrypt c library. It exposes the crypto functions provided by the libmcrypt library in a perl interface 
with a binding code that accounts for null C strings in ciphertext or plain text.

=head1 ATTRIBUTES

=head2 algorithm

contains the name of the algorithm used to decrypt encrypt blocks of data

=head2 mode 

contains the name of the L<block cipher mode of operation|http://en.wikipedia.org/wiki/Block_cipher_modes_of_operation> used to encrypt/decrypt blocks of data.

=head2 key

contains the key for the encryption decryption algorithm

=head2 Initialization Vector 

contains the Initialization for the first block of data in block cipher mode of operation. 

=head1 METHODS

=head2 $self->decrypt($ciphertext)

decrypt blocks of ciphertext 

=head2 $self->encrypt($ciphertext)

encrypt blocks of data

=encoding utf-8

=head1 USAGE

=over

=item *

This module provides a object oriented interface to the libmcrypt library. It uses Mo, a scaled down version of Moose without any data checks to improve speed.

=back

=head1 see ALSO

=over

=item *

L<Mo>

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through github at 
L<https://github.com/shantanubhadoria/crypt-mcrypt/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/shantanubhadoria/crypt-mcrypt>

  git clone git://github.com/shantanubhadoria/crypt-mcrypt.git

=head1 AUTHOR

Shantanu Bhadoria <shantanu at cpan dott org>

=head1 CONTRIBUTORS

=over 4

=item *

Shantanu <shantanu@cpan.org>

=item *

Shantanu Bhadoria <shantanu@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Crypt-MCrypt

0.06 2013-09-12T09:15:48
 - Merge branch 'master' of github.com:shantanubhadoria/Crypt-MCrypt
   [Shantanu]
 - Bump Changes [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.05 2013-09-08T09:04:36
 - Fixed XS syntax bugs [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.04 2013-07-11T00:48:40
 - Added Changes [Shantanu Bhadoria]
 - Added DES ECB Test [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.03 2013-07-09T15:38:54
 - v0.03
   
	   - 2013-07-09 23:38:26 Asia/Singapore [Shantanu Bhadoria]
 - Added README and .gitignore [Shantanu Bhadoria]

0.02 2013-07-09T15:31:31
 - Added README and .gitignore [Shantanu Bhadoria]
 - Modified and REMOVED Default OUR $VERSION line to generate using
   Dist::Zilla, Added Test [Shantanu Bhadoria]
 - Adding XS Files
   triple des cbc decryption test [Shantanu Bhadoria]

0.01 2013-07-09T09:36:15
 - First Interation of the Crypt::MCrypt Module [Shantanu Bhadoria]



inc/MCryptMakeMaker.pm
package inc::MCryptMakeMaker;
use Moose;

extends 'Dist::Zilla::Plugin::MakeMaker::Awesome';

override _build_WriteMakefile_args => sub {
    my ($self) = @_;
    my $template = super();
    our @DIR = qw(.);
    our @LIBS = qw(ltdl mcrypt);
    return +{
        %{ $template },
        LIBS          => [join(' ', map { "-l$_" } @LIBS) . ' '],
        DEFINE        => '',
        INC           => join(' ', map { "-I$_" } @DIR),
    };
};

override _build_MakeFile_PL_template => sub {
    my ($self) = @_;
    my $template = super();

    # Extra code append for Makefile.PL for XS binding 
    $template .= <<'TEMPLATE';
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Crypt::MCrypt',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}
TEMPLATE
    return $template;
};

__PACKAGE__->meta->make_immutable;
1;


lib/Crypt/MCrypt.pm
use strict;
use warnings;
package Crypt::MCrypt;

# PODNAME: Crypt::MCrypt
# ABSTRACT: Perl interface for libmcrypt C library.  
# COPYRIGHT
# VERSION

use 5.010;

# Dependencies
use Mo;
use Carp;


require XSLoader;
XSLoader::load('Crypt::MCrypt', $VERSION);

=encoding utf-8

=attr algorithm

contains the name of the algorithm used to decrypt encrypt blocks of data

=cut

has algorithm => ();

=attr mode 

contains the name of the L<block cipher mode of operation|http://en.wikipedia.org/wiki/Block_cipher_modes_of_operation> used to encrypt/decrypt blocks of data.

=cut

has mode => ();

=attr key

contains the key for the encryption decryption algorithm

=cut 

has key => ();

=attr Initialization Vector 

contains the Initialization for the first block of data in block cipher mode of operation. 

=cut 

has iv => ();

=method $self->decrypt($ciphertext)

decrypt blocks of ciphertext 

=cut

sub decrypt {
    my ($self,$ciphertext) = @_;

    my $length = length($ciphertext);
    my $plain_hex = Crypt::MCrypt::_decrypt($self->algorithm,$self->mode,$ciphertext,$self->key,$length,$self->iv);
    return pack("H*",$plain_hex);
}

=method $self->encrypt($ciphertext)

encrypt blocks of data

=cut

sub encrypt {
    my ($self,$plaintext) = @_;

    my $length = length($plaintext);
    my $cipher_hex = Crypt::MCrypt::_encrypt($self->algorithm,$self->mode,$plaintext,$self->key,$length,$self->iv);
    return pack("H*",$cipher_hex);
}

1;

__END__

=begin wikidoc

= SYNOPSIS

    use Crypt::MCrypt;

    my $iv = pack("H*","0000000000000000");
    my $key = pack("H*","1234567890123456" . "7890123456789012" . "1234567890123456");
    my $cipher_text = pack("H*","E9FF3161EE05ABC9" 
        . "7ea3cacb991318aa" 
        . "585379599b0eaabb" 
        . "c4e474ead1956f47" 
        . "6755f13f1af5235d");
    my $algorithm = "tripledes";
    my $mode = "cbc";
    my $obj = Crypt::MCrypt->new(
        algorithm => $algorithm, 
        mode      => $mode,
        key       => $key, 
        iv        => $iv,
    );
    my $plain_text = $obj->decrypt($cipher_text);
    print "\nPLAIN: $plain_text\n";
    print "\nPLAIN in hex: " . unpack("H*",$plain_text) . "\n";
    $cipher_text = $obj->encrypt($plain_text);
    print "\nCIPHER: $cipher_text\n";
    print "\nCIPHER in hex: " . unpack("H*",$cipher_text) . "\n";

= DESCRIPTION

This is a perl interface to libmcrypt c library. It exposes the crypto functions provided by the libmcrypt library in a perl interface 
with a binding code that accounts for null C strings in ciphertext or plain text.

= USAGE

* This module provides a object oriented interface to the libmcrypt library. It uses Mo, a scaled down version of Moose without any data checks to improve speed.

= see ALSO

* [Mo]

=end wikidoc

=cut
